### PR TITLE
Add serde support to all indicators

### PR DESCRIPTION
**Pull request by dgunay**
Figured I'd just go ahead and open the PR. Adds serde support to all indicators using `serde_derive`. 

Perhaps most importantly I changed the `Next` trait to require the implementation of Serialize and Deserialize:

```rust
pub trait Next<'a, T>: Serialize + Deserialize<'a>
```

There's only a test for one indicator but perhaps it's not even really necessary (since I imagine it would be unlikely to compile if it were broken given how `serde_derive` works).

If you would rather not bloat compilation times and maybe have it behind a feature flag or something that is totally fine, it's up to you. Thanks for maintaining this repo by the way, it was a big help for one of my projects.

### CODE REVIEW

**Comment by greyblake** [comment]
@dgunay thanks for the PR.
If we add serde support, `serde` depedency must be optional and depend on `serde` feature flag (`false` by default) like it's done in other crates (see [chrono](https://github.com/chronotope/chrono#features) for examlpe)

**Comment by dgunay** [comment]
Good idea, I'll look into making it a feature flag.

File: .travis.yml [hunk 1]
```diff
@@ -1,11 +1,15 @@
 language: rust
+cache: cargo # cache dependencies to reduce compilation times
 rust:
   - stable
 install:
   - rustup component add rustfmt
   - rustup component add clippy
 script:
-  - cargo fmt -- --check
+  # - cargo fmt -- --check
```
Comment (greyblake) [inline .travis.yml:9]: Could you please revert this change?

File: .travis.yml [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
   # - cargo clippy -- -D warnings
   - cargo test
   - cargo package
+jobs:
+  allow_failures:
+    - rust: nightly
```
Comment (greyblake) [inline .travis.yml:15]: Could you please remove back this new `jobs` section?

File: Cargo.toml [hunk 3]
```diff
@@ -17,17 +17,22 @@ include = [
     "README.md"
 ]
 
+[features]
+serde_support = ["serde"]
```
Comment (greyblake) [inline Cargo.toml:21]: Could you please rename feature name `serder_support` -> `serde`?

File: src/indicators/average_true_range.rs [hunk 4]
```diff
@@ -69,15 +73,15 @@ impl AverageTrueRange {
     }
 }
 
-impl Next<f64> for AverageTrueRange {
+impl<'a> Next<f64> for AverageTrueRange {
```
Comment (greyblake) [inline src/indicators/average_true_range.rs:76]: Please remove unnecessary liftimes here and in the other places.
(they were removed on purpose a while ago: https://github.com/greyblake/ta-rs/pull/26)

File: src/indicators/bollinger_bands.rs [hunk 5]
```diff
@@ -42,7 +44,8 @@ use crate::{Close, Next, Reset};
 ///
 /// # Links
 ///
-/// * [Bollinger Bands, Wikipedia](https://en.wikipedia.org/wiki/Bollinger_Bands)
+/// ![Bollinger Bands, Wikipedia](https://en.wikipedia.org/wiki/Bollinger_Bands)
```
Comment (greyblake) [inline src/indicators/bollinger_bands.rs:47]: Please revert back this line

File: src/traits.rs [hunk 6]
```diff
@@ -15,6 +18,13 @@ pub trait Reset {
 /// [MACD](indicators/struct.MovingAverageConvergenceDivergence.html) it is `(f64, f64, f64)` since
 /// MACD returns 3 values.
 ///
+#[cfg(feature = "serde_support")]
+pub trait Next<T>: Serialize + DeserializeOwned {
```
Comment (greyblake) [inline src/traits.rs:22]: Please revert changes in `Next` trait. Sorry for  not mentioning this earlier.
The trait should not depend on other traits. If caller/client needs an indicator to be serializable they have to declare requirement for `Serialize + Deserialize` traits explicitly.

As you noticed the traits in this library (Next, Reset) are pretty much thin. It's done so by design with respect to [Interface segregation principle](https://en.wikipedia.org/wiki/Interface_segregation_principle)

It could be that someone wants to implement own indicator and `Next` trait for it, but we should not force them to implement serialization if they don't need it.

**Review by greyblake — COMMENTED** [review]
Thanks for your work.
I've reviewed your changes and left some comments.

File: Cargo.toml (hunk 3 above)
Comment (dgunay) [reply Cargo.toml:21]: Thought this wasn't possible but turns out, it is. Just leaving this here for my reference later: https://users.rust-lang.org/t/features-and-dependencies-cannot-have-the-same-name/47746

**Comment by dgunay** [comment]
One loose end - I don't know if `.travis.yml` will run the tests with the `serde` feature on (or how to do that).

**Comment by greyblake** [comment]
@dgunay Thanks for your update.
One last thing (I hope).. could you please remove all the liftimes that were added in this PR, including those with indicators?

image

**Comment by tirz** [comment]
Hey o/

This PR is pretty cool :) !

> One loose end - I don't know if `.travis.yml` will run the tests with the `serde` feature on (or how to do that).

@dgunay 
You may have a look at [https://github.com/aatxe/irc/blob/develop/.travis.yml](https://github.com/aatxe/irc/blob/develop/.travis.yml)
They set an env value and then run the script like:
```yml
    - &test
      env: FEATURES=default
      script:
        - cargo test --all --no-default-features --features "$FEATURES"
```

**Comment by greyblake** [comment]
@dgunay Thanks for the great work!

**Comment by greyblake** [comment]
@dgunay I just released a new version `0.3.0` with your change.
